### PR TITLE
rpc: Do not use deadlines for rpc connections.

### LINKED ISSUE #4626: Distributed Docker Minio keeps throwing disk errors
I spoke briefly with kp in the Slack channel about this issue.

I have four servers, and one Minio instance on each, running as a four-disk distributed cluster.

It works for a short time, then starts throwing errors a lot, like:

`msg="Unable to fetch disk info for &cmd.networkStorage{rpcClient:(*cmd.AuthRPCClient)(0xc4202a9760)}" cause="disk not found" source="[xl-v1.go:199:getDisksInfo()]"`

Running `mc admin info` says that three of the four nodes see three disks offline each; but after some seconds, it goes back to saying all disks online. But attempting to make a bucket fails if run again.

This is the setup. Assume you have four nodes, called m1, m2, m3, m4, then run this on each one:

```
mkdir -p /opt/minio/export
docker run --rm --hostname={{ ansible_fqdn}} \
    -p 9000:9000 \
    -v /opt/minio/export:/export \
    -e MINIO_ACCESS_KEY=something \
    -e MINIO_SECRET_KEY=somethingsecret \
    minio/minio server --address :9000 \
    http://m1:9000/export \
    http://m2:9000/export \
    http://m3:9000/export \
    http://m4:9000/export

```

Configure `mc` to talk to it, and just start creating buckets and adding files to them, removing files, and removing the bucket, with a delay. I guess you have test scripts like these that are more powerful than mine! (which is essentially a few lines, a counter, and a sleep, so I end up with `mc mb test/test$n` etc)

I find it starts failing quite quickly, especially if I run `mc admin heal test` a few times.

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
Disable deadlines on RPC connections.

## Motivation and Context
Fixes #4626

## How Has This Been Tested?
Manually using a distributed setup.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by deekoder** [comment]
As per @harshavardhana blocked on this until #4636 is merged.

**Comment by harshavardhana** [comment]
This PR is updated and its not blocked anymore.. ready for reviews.

**Comment by krishnasrinivas** [comment]
@harshavardhana 
compile failed
```
krishna@escape:~/gocode/src/github.com/minio/minio$ go install
# github.com/minio/minio/cmd
cmd/rpc-server.go:77: undefined: "github.com/minio/minio/pkg/http".BuffConn
krishna@escape:~/gocode/src/github.com/minio/minio$ 

```

File: cmd/rpc-server.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+	if req.Method != http.MethodConnect {
+		w.Header().Set("Content-Type", "text/plain; charset=utf-8")
+		w.WriteHeader(http.StatusMethodNotAllowed)
+		io.WriteString(w, fmt.Sprintf("405 must %s\n", http.MethodConnect))
+		return
+	}
+
+	conn, _, err := w.(http.Hijacker).Hijack()
+	if err != nil {
+		log.Printf("rpc hijacking %s: %s\n", req.RemoteAddr, err.Error())
```
Comment (krishnasrinivas) [inline cmd/rpc-server.go]: log.Printf should be removed?

**Comment by krishnasrinivas** [comment]
The problem does not seem to be fixed @harshavardhana 

if I ctrl-C in distributed mode, I should see :
```
^CExiting on signal interrupt

ERRO[0044] Unable to shutdown http server                cause=timed out. some connections are still active. doing abnormal shutdown source=[signals.go:42:handleSignals.func2()]
```

but I don't see it when I ctrl-C

File: cmd/rpc-server.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+	"io"
+	"net"
+	"net/http"
+	"net/rpc"
+
+	http2 "github.com/minio/minio/pkg/http"
+)
+
+// Can connect to RPC service using HTTP CONNECT to rpcPath.
+var connected = "200 Connected to Go RPC"
```
Comment (balamurugana) [inline cmd/rpc-server.go]: This needs to be `const` and name could be `rpcConnectReply`

**Comment by harshavardhana** [comment]
Done with all the final changes @balamurugana @donatello @krishnasrinivas . RPC connections are holding up as ESTABLISHED. Any questions let me know we can discuss further.

File: cmd/rpc-server.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+package cmd
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/rpc"
+
+	http2 "github.com/minio/minio/pkg/http"
```
Comment (balamurugana) [inline cmd/rpc-server.go]: you can name it as `miniohttp`

File: cmd/rpc-server.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+)
+
+// Can connect to RPC service using HTTP CONNECT to rpcPath.
+const rpcConnectReply = "200 Connected to Go RPC"
+
+func (server *rpcServer) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	if req.Method != http.MethodConnect {
+		w.Header().Set("Content-Type", "text/plain; charset=utf-8")
+		w.WriteHeader(http.StatusMethodNotAllowed)
+		io.WriteString(w, fmt.Sprintf("405 must %s\n", http.MethodConnect))
```
Comment (balamurugana) [inline cmd/rpc-server.go]: I think we don't need to send a body. If we do we can remove header setting also.

File: cmd/rpc-server.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+	}
+
+	conn, _, err := w.(http.Hijacker).Hijack()
+	if err != nil {
+		errorIf(err, "rpc hijacking: %s", req.RemoteAddr)
+		return
+	}
+
+	// Overrides Read/Write deadlines if any.
+	conn = newRPCConn(conn)
```
Comment (balamurugana) [inline cmd/rpc-server.go]: this function is not required. you could directly use it like
```go
bufConn, ok := conn.(*miniohttp.BufConn)
if ok {
    bufConn.DisableTimeout()
    conn = bufConn
}
```

File: cmd/rpc-server.go [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+	io.WriteString(conn, "HTTP/1.0 "+rpcConnectReply+"\n\n")
```
Comment (balamurugana) [inline cmd/rpc-server.go]: You could keep the whole string ie `HTTP/1.0 200 Connected to Go RPC\r\n` in `rpcConnectReply`

File: cmd/rpc-server.go [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+	server.ServeConn(conn)
+}
+
+type rpcServer struct {
+	*rpc.Server
```
Comment (balamurugana) [inline cmd/rpc-server.go]: If `rpcServer` extends `rpc.Server`, then `newRPCServer()` is simply `return &rpcServer{}`

Comment (harshavardhana) [reply cmd/rpc-server.go]: we need to do rpc.NewServer() we can't do `&rpcServer{}` it won't allocate otherwise.

Comment (balamurugana) [reply cmd/rpc-server.go]: Got it.

File: cmd/rpc-server.go [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+	"io"
+	"net/http"
+	"net/rpc"
+
+	miniohttp "github.com/minio/minio/pkg/http"
+)
+
+func (server *rpcServer) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	if req.Method != http.MethodConnect {
+		w.Header().Set("Content-Type", "text/plain; charset=utf-8")
```
Comment (balamurugana) [inline cmd/rpc-server.go]: Header setting is not useful anymore.

**Comment by harshavardhana** [comment]
The bug reported #4677 is also fixed by this PR. Since this PR test case was blocking on it..

File: cmd/rpc-server_test.go [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+		t.Fatal(err)
+	}
+
+	if resp.StatusCode != http.StatusMethodNotAllowed {
+		t.Errorf("Expected %d, got %d", http.StatusMethodNotAllowed, resp.StatusCode)
+	}
+}
+
+/*
+func TestHTTPRPC(t *testing.T) {
```
Comment (krishnasrinivas) [inline cmd/rpc-server_test.go]: this should be removed?

File: cmd/rpc-server.go [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+	bufConn, ok := conn.(*miniohttp.BufConn)
+	if ok {
+		bufConn.RemoveTimeout()
+		conn = bufConn
+	}
+
+	// Can connect to RPC service using HTTP CONNECT to rpcPath.
+	io.WriteString(conn, "HTTP/1.0 200 Connected to Go RPC\n\n")
+	server.ServeConn(conn)
+}
```
Comment (krishnasrinivas) [inline cmd/rpc-server.go:51]: instead you can just do this:
```
func (server *rpcServer) ServeHTTP(w http.ResponseWriter, req *http.Request) {                                                                                                  
        conn, _, err := w.(http.Hijacker).Hijack()                                                                                                                              
        if err != nil {                                                                                                                                                         
                errorIf(err, "rpc hijacking failed for: %s", req.RemoteAddr)                                                                                                    
                return                                                                                                                                                          
        }                                                                                                                                                                       
                                                                                                                                                                                
        // Overrides Read/Write deadlines if any.                                                                                                                               
        bufConn, ok := conn.(*miniohttp.BufConn)                                                                                                                                
        if ok {                                                                                                                                                                 
                bufConn.RemoveTimeout()                                                                                                                                         
        }                                                                                                                                                                       
                                                                                                                                                                                
        server.Server.ServeHTTP(w, req)                                                                                                                                         
}
```

Comment (harshavardhana) [reply cmd/rpc-server.go:51]: Once hijacked you cant put it back.. you have to now manage the net.Conn

File: cmd/rpc-server_test.go (hunk 9 above)
Comment (harshavardhana) [reply cmd/rpc-server_test.go]: You are looking at an old PR.

File: cmd/rpc-server.go (hunk 10 above)
Comment (krishnasrinivas) [reply cmd/rpc-server.go:51]: ok

**Review by balamurugana — APPROVED** [review]

**Review by krishnasrinivas — APPROVED** [review]

**Pull request merged**